esx_service/utils/auth_data.py

```diff
@@ -417,14 +420,223 @@ def remove_datastore_access_privileges(self, conn, datastore_url):
+
+    # this is in-memory shared DB URL
+    MEMDB_NAME = "file::memory:?cache=shared"
+
+    # - How often to refresh memdb from on-disk DB.  Default 15 min.
+    REFRESH_SEC = int(os.getenv("VMDKOPS_MEMDB_REFRESH_SEC", default=900))
+
+
+    def __init__(self):
+        # Init connection to in-memory DB.
+        # We keep one connection open to assure the DB is not garbage collected too early
+        # We refresh it every MEMDB_REFRESH_SEC i. When we refresh, we first drain all in-flight
+        # AuthernticatoinDataManagers to make sure nobody is using the copy.
+        #
+        # 'None' means the feature is disabled
+        self._memdb_con = None
+
+        # Init Barrier
+        # We need to avoid refresh (which could be scheduled by Timer on a different thread)
+        # from messing DB content for auth. mgrs in flight. So  we acquire a barrier,
+        # and wait until all auth.managers in flight land.
+        self._barrier = threading.Condition()
+        self._in_flight = 0
+
+        # Init timer for periodic refreshes
+        self._timer = None
+
+    def __del__(self):
+        pass
+
+    def _drain_and_close(self):
+        """
+        Drains Auth. managers in fight and closes last memdb connection.
+        *** EXPECTS to be run under _barrier lock.
+        """
+        logging.debug("_Drain and close")
+        # drain the in-flight users
+        self._memdb_con.close()
+        self._memdb_con = None
+        while self._in_flight > 0:
+            logging.debug("Wating for inflight %d", self._in_flight)
+            self._barrier.wait()
+        logging.debug("DONE")
+
+
+    def enable(self):
+        """
+        Enables in-memory ReadOnly cache DB for auth DB.
+        This includes opening on-disk DB and populating in-memory DB from it.
+        Note this will force the connection to treat the DB as read only.
+        """
+        logging.debug("memdb_enable")
+        if not self.is_enabled():
+            logging.debug("memdb_enable Starting refresh")
+            # No need to drain anything as nobody was using the cache so far
+            # Proceed to creation os a new cache
+            self._memdb_refresh_periodically(frequency_sec=self.__class__.REFRESH_SEC)
+
+    def disable(self):
+        """
+        Disables the DB cache and closes the connection. The  DB will be deleted when all
+        AuthorizationDataManagers using this cache land and close their connection.
+        """
+        logging.debug("memdb_disable")
+        with self._barrier:
+            if self.is_enabled():
+                self._timer.cancel()
+                self._drain_and_close()
+        logging.debug("Disabled")
+
+
+    def is_enabled(self):
+        'True if memdb cache is enabled'
+        with self._barrier:
+            ret = self._memdb_con is None
+        return ret
+
+    @staticmethod
+    def _make_connection_readonly(con):
+        con.executescript("PRAGMA query_only = 1; PRAGMA read_uncommitted = 1;")
```

Comment:
FWIW since *all* connections to the memdb after it's created are readonly (query_only = 1) there should be no transactions to abort :-)

So I've added out of sheer paranoia, in case sqlite3 has some internal checks we can skip.